File: tests/cpp/test_Gneiting.cpp
/******************************************************************************/
/*                                                                            */
/*                            gstlearn C++ Library                            */
/*                                                                            */
/* Copyright (c) (2023) MINES Paris / ARMINES                                 */
/* Authors: gstlearn Team                                                     */
/* Website: https://gstlearn.org                                              */
/* License: BSD 3-clause                                                      */
/*                                                                            */
/******************************************************************************/
#include "Basic/OptCustom.hpp"
#include "Basic/VectorNumT.hpp"
#include "Covariances/CorAniso.hpp"
#include "Covariances/CorGaussianMixture.hpp"
#include "Covariances/CorGneiting.hpp"
#include "Covariances/CovContext.hpp"
#include "Db/Db.hpp"
#include "Db/DbGrid.hpp"
#include "Db/DbStringFormat.hpp"
#include "Estimation/CalcKriging.hpp"
#include "Model/Model.hpp"
#include "Neigh/NeighUnique.hpp"
#include "Space/ASpaceObject.hpp"
#include "Space/SpaceComposite.hpp"
#include "Space/SpacePoint.hpp"
#include "Space/SpaceRN.hpp"

using namespace gstlrn;
/**
 * This file is meant to test Kriging with Gneiting Model
 */
int main(int argc, char* argv[])
{
  OptCustom::define("ompthreads", 1);
  std::stringstream sfn;
  sfn << gslBaseName(__FILE__) << ".out";
  StdoutRedirect sr(sfn.str(), argc, argv);
  auto space1d = SpaceRN::create(1);
  auto space2d = SpaceRN::create(2);
  auto sp      = SpaceComposite::create({space2d, space1d});
  sp->display();
  setDefaultSpace(sp);

  // creating the time trace covariance
  CovContext ctxt1d(1, space1d);
  double alpha  = 1.0;
  double beta   = 1.0;
  double scaleT = 5.3;
  auto* corT    = new CorAniso(CovContext(1, 1), ECov::CAUCHY_GEN);
  corT->setParam(alpha, 0);        // alpha in (0,2]
  corT->setParam(beta * 2 / 2, 1); // beta*d/2 with beta in (0,1]
  corT->setScaleDim(0, scaleT);

  // creating the space trace covariance
  CovContext ctxt2d(1, space2d);
  VectorDouble scales = {2., 3.};
  VectorDouble params = {0.5};
  VectorDouble kappas = {1.0};
  VectorDouble angles = {30.0, 0.0};
  auto* corS          = CorGaussianMixture::create(
    ctxt2d,
    ECov::MATERN,
    params,
    kappas,
    scales,
    angles,
    false);
  double sep = 1.;
  CorGneiting corGneiting(corS, corT, sep);
  message("Space dimension of Gneiting Covariance = %d\n", corGneiting.getNDim());

  // Testing the covariance calculation between two points
  VectorDouble coords1 = {12., 3., 1.};
  VectorDouble coords2 = {4., 5., 2.};
  SpacePoint p1(sp);
  SpacePoint p2(sp);
  p1.setCoords(coords1);
  p2.setCoords(coords2);
  double cres = corGneiting.evalCov(p1, p2);
  std::cout << "Value of Gneiting (by Covariance) = " << cres << std::endl;

  // Create the Data Base
  Id ndim  = 3;
  Id ndat  = 10;
  Id nvar  = 1;
  Db* data = Db::createFillRandom(ndat, ndim, nvar);

  // Create the Target
  VectorInt nx    = {5, 5, 2};
  VectorDouble dx = {1. / nx[0], 1. / nx[1], 1. / nx[2]};
  DbGrid* grid    = DbGrid::create(nx, dx);

  // Create the Model
  auto* model = new ModelGeneric();
  model->setCov(&corGneiting);
  model->evalCov(p1, p2);
  message("Model dimension = %d\n", model->getNDim());
  std::cout << "Value of Gneiting (by Model) = " << cres << std::endl;

  // Create the Unique neighborhood
  NeighUnique* neigh = NeighUnique::create(false, sp);
  // Launch Kriging
  (void)kriging(data, grid, model, neigh);

  // Display a summary of the results
  DbStringFormat dbfmtKriging(FLAG_STATS);
  grid->display(&dbfmtKriging);

  delete corT;
  delete corS;
  delete data;
  delete grid;
  delete neigh;
  delete model;

  return (0);
}

File: tests/py/test_Gneiting.py
import gstlearn as gl
import numpy as np

# creating the space-time
space1D = gl.SpaceRN.create(1)
space2D = gl.SpaceRN.create(2)

# creating the time trace covariance
ctxt1D = gl.CovContext(1, space1D)
alpha = 1.0
beta = 1.0
params = [alpha, beta * 2 / 2]
scaleT = 5.3
angleT = 0.0
corT = gl.CorAniso.create(ctxt1D, gl.ECov.CAUCHY_GEN, params, scaleT, angleT, False)

# creating the space trace covariance
ctxt2D = gl.CovContext(1, space2D)
nu = [0.5]
kappa = [1.0]
scales = [2.0, 3.0]
angles = [0, 0]
corS = gl.CorGaussianMixture.create(
    gl.ECov.MATERN, ctxt2D, nu, kappa, scales, angles, False
)

sep = 1
gneiting = gl.CorGneiting(corS, corT, sep)
print("Space dimension of Gneiting Covariance = " + str(gneiting.getNDim()))

coords1 = [12, 3, 1]
coords2 = [4, 5, 2]
p1 = gl.SpacePoint(gneiting.getSpace())
p2 = gl.SpacePoint(gneiting.getSpace())
p1.setCoords(coords1)
p2.setCoords(coords2)
cres = gneiting.evalCov(p1, p2)
print("Gneiting eval " + str(cres))

p1_0 = p1.spacePointOnSubspace(0)
print("Displaying the first two coordinates (spatial) of the first space point")
p1_0.display()

p1_1 = p1.spacePointOnSubspace(1)
print("Displaying the last coordinate (temporal) of the first space point")

p1_1.display()

p2_0 = p2.spacePointOnSubspace(0)
print("Displaying the first two coordinates (spatial) of the second space point")
p2_0.display()

p2_1 = p2.spacePointOnSubspace(1)
print("Displaying the last coordinate (temporal) of the second space point")

p2_1.display()

# temporal trace covariance
covt_gstl = corT.evalCov(p1_1, p2_1)
covt_calc = 1 / (
    (1 + (np.abs(coords1[2] - coords2[2]) / scaleT) ** params[0]) ** params[1]
)
print(
    "Difference for temporal covariance = "
    + str(np.round(np.abs(covt_gstl - covt_calc), 6))
)

# spatial trace covariance
delta = [np.abs(coords1[0] - coords2[0]), np.abs(coords1[1] - coords2[1])]
covs_gstl = corS.evalCov(p1_0, p2_0)
covs_calc = np.exp(-np.sqrt((delta[0] / scales[0]) ** 2 + (delta[1] / scales[1]) ** 2))
print(
    "Difference for spatial covariance = "
    + str(np.round(np.abs(covs_gstl - covs_calc), 6))
)

# gneiting covariance
al = sep / 2
scale = covt_gstl**al
covspatCopy = gl.CorGaussianMixture(corS)
covspatCopy.setScaleGneiting(scale)
covs = covspatCopy.evalCov(p1_0, p2_0)

print(
    "Difference for Gneiting covariance = "
    + str(np.round(np.abs(cres - covs * covt_gstl), 6))
)
